spec/mem/tree.md

```diff
@@ -0,0 +1,564 @@
+# MiniRust Tree Borrows
+
+To check the details on Tree Borrows
+
+1. [From Stacks to Trees: A new aliasing model for Rust](https://www.ralfj.de/blog/2023/06/02/tree-borrows.html)
+1. [Neven's posts on Tree Borrows](https://perso.crans.org/vanille/treebor)
+
+## Data structures
+
+In addition to the `BasicMemory`, `TreeBorrowsMemory` also track an ID for each pointer, called a *Borrow Tag*.
+
+```rust
+pub struct BorTag(Int);
+pub type TreeBorrowsProvenance = (BorTag, AllocId);
+```
+
+```rust
+pub struct TreeBorrowsMemory<T: Target> {
+    allocations: List<TreeBorrowsAllocation>,
+    /// Next unused borrow tag.
+    next_tag: Int,
+    // FIXME: specr should add this automatically
+    _phantom: std::marker::PhantomData<T>,
+}
+
+impl<T: Target> Memory for TreeBorrowsMemory<T> {
+    type Provenance = TreeBorrowsProvenance;
+    type T = T;
+
+    fn new() -> Self {
+        Self { allocations: List::new(), next_tag: Int::ZERO, _phantom: std::marker::PhantomData }
+    }
+}
+```
+
+We track each tag's permissions on specific memory locations (bytes)
+
+```rust
+pub enum AccessKind {
+    Read, 
+    Write,
+}
+
+pub enum Permission {
+    /// Represents a two-phase borrow during its reservation phase
+    Reserved, 
+    /// Represents an activated (written to) mutable reference
+    Active,
+    /// Represents a shared (immutable) reference
+    Frozen,
+    /// Represents a dead reference
+    Disabled, 
+}
+```
+
+We create one *tree* for tracking all pointers in each allocation.
+
+```rust
+pub struct TreeBorrowsAllocation {
+    /// The data stored in this allocation.
+    data: List<AbstractByte<TreeBorrowsProvenance>>,
+    /// The address where this allocation starts.
+    /// This is never 0, and `addr + data.len()` fits into a `usize`.
+    addr: Address,
+    /// The alignment that was requested for this allocation.
+    /// `addr` will be a multiple of this.
+    align: Align,
+    /// The kind of this allocation.
+    kind: AllocationKind,
+    /// Whether this allocation is still live.
+    live: bool,
```

Comment:
Instead of repeating all the fields of a basic `Allocation`, would it make sense to just re-use the basic `Allocation` type here? And then also re-use some of the methods from basic.md for memory access etc?